powerauth-data-adapter/src/main/java/io/getlime/security/powerauth/lib/dataadapter/controller/AuthenticationController.java

```diff
@@ -58,7 +63,18 @@ public AuthenticationController(DataAdapter dataAdapter) {
     @RequestMapping(value = "/authenticate", method = RequestMethod.POST)
     public @ResponseBody ObjectResponse<AuthenticationResponse> authenticate(@RequestBody ObjectRequest<AuthenticationRequest> request) throws MethodArgumentNotValidException, AuthenticationFailedException {
         AuthenticationRequest authenticationRequest = request.getRequestObject();
-        AuthenticationResponse response = dataAdapter.authenticateUser(authenticationRequest);
+        // input validation is handled by AuthenticationRequestValidator
+        // validation is invoked manually because of the generified Request object
+        AuthenticationRequestValidator validator = new AuthenticationRequestValidator();
+        BeanPropertyBindingResult result = new BeanPropertyBindingResult(authenticationRequest, "authenticationRequest");
+        ValidationUtils.invokeValidator(validator, authenticationRequest, result);
+        if (result.hasErrors()) {
+            // getEnclosingMethod() on new object returns a reference to current method
+            MethodParameter methodParam = new MethodParameter(new Object(){}.getClass().getEnclosingMethod(),0);
+            throw new MethodArgumentNotValidException(methodParam, result);
+        }
+        dataAdapter.authenticateUser(authenticationRequest.getUsername(), authenticationRequest.getPassword());
+        AuthenticationResponse response = new AuthenticationResponse(authenticationRequest.getUsername());
```

Comment:
This is the issue - "username" is not "userId". Generally, these values differ. While username in a typical bank is either a client number, e-mail or username, it always translates to canonical user ID...